server/src-lib/Hasura/RQL/DDL/RemoteSchema.hs

```diff
@@ -156,3 +167,130 @@ fetchRemoteSchemas =
      |] () True
   where
     fromRow (n, Q.AltJ def, comm) = AddRemoteSchemaQuery n def comm
+
+runAddRemoteSchemaPermissions
+  :: ( QErrM m, UserInfoM m
+     , CacheRWM m, MonadTx m
+     )
+  => RemoteSchemaPermissions -> m EncJSON
+runAddRemoteSchemaPermissions q = do
+  adminOnly
+  newRSCtx <- runAddRemoteSchemaPermissionsP1 q
+  runAddRemoteSchemaPermissionsP2 q newRSCtx
+  pure successMsg
+
+runAddRemoteSchemaPermissionsP1
+  :: (QErrM m, CacheRM m) => RemoteSchemaPermissions -> m RemoteSchemaCtx
+runAddRemoteSchemaPermissionsP1 remoteSchemaPermission = do
+  validateRemoteSchemaPermissions remoteSchemaPermission
+
+runAddRemoteSchemaPermissionsP2
+  :: (QErrM m, CacheRWM m, MonadTx m) => RemoteSchemaPermissions -> RemoteSchemaCtx -> m ()
+runAddRemoteSchemaPermissionsP2 rsPerm rsCtx = do
+  runAddRemoteSchemaPermissionsP2Setup rsPerm rsCtx
+  liftTx $ addRemoteSchemaPermissionsToCatalog rsPerm
+
+runAddRemoteSchemaPermissionsP2Setup
+  :: (QErrM m, CacheRWM m) => RemoteSchemaPermissions -> RemoteSchemaCtx -> m ()
+runAddRemoteSchemaPermissionsP2Setup RemoteSchemaPermissions{..} rsCtx = do
+  sc <- askSchemaCache
+  let rmSchemaWithRoles = scRemoteSchemasWithRole sc
+      newSchemasWithRoles = Map.insert (rsPermRole, rsPermRemoteSchema) rsCtx rmSchemaWithRoles
+  writeSchemaCache sc { scRemoteSchemasWithRole = newSchemasWithRoles }
+
+validateRemoteSchemaPermissions :: (QErrM m, CacheRM m) => RemoteSchemaPermissions -> m RemoteSchemaCtx
+validateRemoteSchemaPermissions remoteSchemaPerm = do
+  sc <- askSchemaCache
+  case Map.lookup
+       (rsPermRemoteSchema remoteSchemaPerm)
+       (scRemoteSchemas sc) of
+    Nothing  -> throw400 RemoteSchemaError "No such remote schema"
+    Just remoteSchemaCtx -> do
+      newTypes <- do
+        checkUniqueTypes (rsPermDefinition remoteSchemaPerm)
+        validateTypes (rsPermDefinition remoteSchemaPerm) (GC._rgTypes $ rscGCtx remoteSchemaCtx)
+      newGCtx <- updateRemoteGCtxFromTypes newTypes (rscGCtx remoteSchemaCtx)
+      pure $ remoteSchemaCtx { rscGCtx = newGCtx }
+
+checkUniqueTypes :: (QErrM m) => [RemoteTypePerm] -> m ()
+checkUniqueTypes typePerms = do
+  let types = map rtpType typePerms
+      nubbedTypes = List.nub types
+  if types /= nubbedTypes
```

Comment:
There's `duplicates` function in `Hasura.Server.Utils` module. Does it fit here?

Reply:
It probably does. But why use a function that is already available in a core lib?